src/s3/builders/list_objects.rs

```diff
@@ -0,0 +1,737 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Argument builders for ListObject APIs.
+
+use async_trait::async_trait;
+use futures_util::{stream as futures_stream, Stream, StreamExt};
+use http::Method;
+
+use crate::s3::{
+    client::Client,
+    error::Error,
+    response::{
+        ListObjectVersionsResponse, ListObjectsResponse, ListObjectsV1Response,
+        ListObjectsV2Response,
+    },
+    types::{S3Api, S3Request, ToS3Request, ToStream},
+    utils::{check_bucket_name, merge, Multimap},
+};
+
+fn add_common_list_objects_query_params(
+    query_params: &mut Multimap,
+    delimiter: Option<&str>,
+    disable_url_encoding: bool,
+    max_keys: Option<u16>,
+    prefix: Option<&str>,
+) {
+    query_params.insert(
+        String::from("delimiter"),
+        delimiter.unwrap_or("").to_string(),
+    );
+    query_params.insert(
+        String::from("max-keys"),
+        max_keys.unwrap_or(1000).to_string(),
+    );
+    query_params.insert(String::from("prefix"), prefix.unwrap_or("").to_string());
+    if !disable_url_encoding {
+        query_params.insert(String::from("encoding-type"), String::from("url"));
+    }
+}
+
+/// Argument builder for ListObjectsV1 S3 API, created by
+/// [list_objects_v1()](crate::s3::client::Client::list_objects_v1).
+#[derive(Clone, Debug, Default)]
+pub struct ListObjectsV1 {
+    client: Option<Client>,
+
+    extra_headers: Option<Multimap>,
+    extra_query_params: Option<Multimap>,
+    region: Option<String>,
+    bucket: String,
+    delimiter: Option<String>,
+    disable_url_encoding: bool,
+    max_keys: Option<u16>,
+    prefix: Option<String>,
+    marker: Option<String>,
+}
+
+#[async_trait]
+impl ToStream for ListObjectsV1 {
+    type Item = ListObjectsV1Response;
+
+    async fn to_stream(self) -> Box<dyn Stream<Item = Result<Self::Item, Error>> + Unpin + Send> {
```

Comment:
This is not a purpose of low level S3 method. We would need to return single parsed response for single request. This defeats the purpose i.e. custom list implementation is not possible.

Reply:
@balamurugana Please see the other comments - the `ToStream` instance is needed for the higher level call's implementation. For custom implementation one can use `S3Api::send()` for each other the lower level APIs.